webapp-django/crashstats/base/ga.py

```diff
@@ -0,0 +1,130 @@
+
+from django.contrib.sites.requests import RequestSite
+from django.conf import settings
+
+logger = logging.getLogger('crashstats:ga')
+
+# Uncomment for local debugging. This will print all logging on stdout.
+# logger.setLevel(logging.DEBUG)
+# import sys
+# logger.addHandler(logging.StreamHandler(sys.stdout))
+
+
+def track_api_pageview(
+    request,
+    page_title=None,
+    data_source='api',
+    client_id=None,
+    **headers
+):
+    """Convenient wrapper function geared for the API calls. This way
+    the page title is automatically guessed to something sensible."""
+    page_title = page_title or 'API ({})'.format(request.path)
+    track_pageview(
+        request,
+        page_title=page_title,
+        data_source=data_source,
+        client_id=client_id,
+        **headers
+    )
+
+
+def track_pageview(
+    request,
+    page_title,
+    data_source='web',
+    client_id=None,
+    **headers
+):
+    """Trigger a HTTP POST to Google Analytics about a particular page being
+    viewed. Most paramters are being picked out of the request object.
+
+    :arg request: The Django request object.
+    :arg page_title: A string to act as a title for the page. Not relevant
+    or particularly needed for API requests.
+    :arg data_source: A string like 'web' or 'api'. See documentation.
+    :arg client_id: Some string to identify the client.
+
+    """
+    if not settings.GOOGLE_ANALYTICS_ID:
+        logging.debug('GOOGLE_ANALYTICS_ID not set up. No pageview tracking.')
+        return
+
+    domain = settings.GOOGLE_ANALYTICS_DOMAIN
+    if not domain or domain == 'auto':
+        domain = RequestSite(request).domain
+
+    params = {}
+    params['v'] = 1  # version
+    params['tid'] = settings.GOOGLE_ANALYTICS_ID  # Tracking ID / Property ID
+    params['dh'] = domain
+    params['t'] = 'pageview'
+    params['ds'] = data_source
+
+    # Here's what the documentation says on
+    # https://developers.google.com/analytics/devguides/collection/protocol\
+    # /v1/parameters#cid
+    #
+    #   Required for all hit types.
+    #
+    #   This anonymously identifies a particular user, device, or browser
+    #   instance. For the web, this is generally stored as a first-party
+    #   cookie with a two-year expiration. For mobile apps, this is randomly
+    #   generated for each particular instance of an application install.
+    #   The value of this field should be a random UUID (version 4) as
+    #   described in http://www.ietf.org/rfc/rfc4122.txt
+    params['cid'] = client_id or uuid.uuid4().hex
+
+    if not request.user.is_anonymous():
+        params['uid'] = str(request.user.id)
+
+    params['dp'] = request.path  # Page
+    if request.META.get('QUERY_STRING'):
+        params['dp'] += '?{}'.format(request.META.get('QUERY_STRING'))
+
+    params['dt'] = page_title
+
+    if request.META.get('REMOTE_ADDR'):
+        ip = request.META['REMOTE_ADDR']
+        if ip != '127.0.0.1':
+            params['uip'] = ip
```

Comment:
One thing this does is it subverts anyone's preferences for being tracked. If I don't want to be tracked, then I can prevent GA from loading in my browser and doing things. However, since we're doing this server-side, even if I do that, I'm still getting tracked.

Thus I'm not wildly excited about using peoples' ip addresses here.

With Input, I "faked" the id by using the request object's id. In that way, it's not tied to someone in a way that's identifying.

Do you have something similar you can use? A Django session id, perhaps?